pipeline/psth.py

```diff
@@ -19,129 +22,229 @@
+            },
+            {
+                'trial_condition_desc': 'good_noearlylick_left_hit',
+                'trial_condition_func': '_get_trials_no_stim',
+                'trial_condition_arg': {
+                    'task': 'audio delay',
+                    'task_protocol': 1,
+                    'outcome': 'hit',
+                    'early_lick': 'no early',
+                    'trial_instruction': 'left'}
+            },
+            {
+                'trial_condition_desc': 'good_noearlylick_right_hit',
+                'trial_condition_func': '_get_trials_no_stim',
+                'trial_condition_arg': {
+                    'task': 'audio delay',
+                    'task_protocol': 1,
+                    'outcome': 'hit',
+                    'early_lick': 'no early',
+                    'trial_instruction': 'right'}
+            },
+            {
+                'trial_condition_desc': 'good_noearlylick_left_miss',
+                'trial_condition_func': '_get_trials_no_stim',
+                'trial_condition_arg': {
+                    'task': 'audio delay',
+                    'task_protocol': 1,
+                    'outcome': 'miss',
+                    'early_lick': 'no early',
+                    'trial_instruction': 'left'}
+            },
+            {
+                'trial_condition_desc': 'good_noearlylick_right_miss',
+                'trial_condition_func': '_get_trials_no_stim',
+                'trial_condition_arg': {
+                    'task': 'audio delay',
+                    'task_protocol': 1,
+                    'outcome': 'miss',
+                    'early_lick': 'no early',
+                    'trial_instruction': 'right'}
+            },
+        )
+
+        return ({**d, 'trial_condition_id': key_hash(d['trial_condition_arg'])}
+                for d in contents_data)
+
+    @classmethod
+    def get_trials(cls, trial_condition_desc):
+        return cls.get_func({'trial_condition_desc': trial_condition_desc})()
+
+    @classmethod
+    def get_func(cls, key):
+        self = cls()
+
+        func, args = (self & key).fetch1(
+            'trial_condition_func', 'trial_condition_arg')
+
+        return partial(dict(getmembers(cls))[func], **args)
+
+    @classmethod
+    def _get_trials_no_stim(cls, task=None, task_protocol=None, outcome=None,
+                            early_lick=None, trial_instruction=None):
+
+        log.debug('_get_trials_no_stim', locals())
+
+        return ((experiment.BehaviorTrial
+                 & {'task': task}
+                 & {'trial_instruction': trial_instruction}
+                 & {'early_lick': early_lick}
+                 & {'outcome': outcome}) - experiment.PhotostimEvent)
+
+    @classmethod
+    def _get_trials_stim(cls, task=None, task_protocol=None, outcome=None,
+                         early_lick=None, trial_instruction=None):
+
+        log.debug('_get_trials_stim', locals())
+
+        return ((experiment.BehaviorTrial
```

Comment:
Does restricting by `None` behave well? If not, perform some exclusion.

Reply:
no - need to rework (see 'FIXME' on line 59; will adjust before completion)

Reply:
handled now (simply applying **kwargs rather than explicit processing)